src/core.rs

```diff
@@ -672,21 +682,34 @@ impl<T> Webauthn<T> {
         &self,
         creds: Vec<Credential>,
         policy: Option<UserVerificationPolicy>,
-        extensions: Option<JSONExtensions>
+        extensions: Option<JSONExtensions>,
     ) -> Result<(RequestChallengeResponse, AuthenticationState), WebauthnError>
     where
         T: WebauthnConfig,
     {
         let chal = self.generate_challenge();
 
-        let mut policy = policy.unwrap_or(UserVerificationPolicy::Preferred_DO_NOT_USE);
-
-        let any_creds_are_verified = creds.iter().fold(false, |cur, next| cur || next.verified );
+        // the credential is verified and was registered with verification required
+        let need_user_verification = creds.iter().fold(false, |cur, next| {
```

Comment:
Still don't like this section. I think in the suggestion I made we should have two interfaces.

For generate_challenge_authenticate() it takes a Vec of credentials, and *no* policy. It has the same consistent vs inconsistent error with mixed type.

We should also have a second interface, generate_challenge_authenticate_single, which takes a single credential, and allows the policy to be over-ridden. 

That way we satisfy both your needs to have the policy changeable for a single credential (meaning the implementation has an out-of-band ability to indicate what credential should be used in this authentication, inline with the unwritten intent of the spec), and my desire that in "mixed" scenarios the policy is consistent and enforced with out "changing" the policy underneath. 

Otherwise I think all the rest looks really good!

Reply:
I still think there's too much inner logic that will easily be mis-used. For example, I just ran into a 3rd case: an empty list of credentials (i.e. a user-less operation mode). In this case it **should** take a policy.

Probably you're right that it's best to have separate authenticate methods for each type of use case. There should be *two* methods that take multiple credentials -- one that takes a policy and one that doesn't. If it doesn't take a policy then it infers from the credentials existing values.

Reply:
What would be the use case with an empty credential list? I seem to recall there is a part of the spec where you don't send a list of "allowed creds" but the server still needs the credentials available to understand how to do the authenticaiton. So there would have to be a seperate case here.

But yes, different auth methods I think is the right way here because that way the caller has a strongly typed interface, and we can document each one with exactly when and how you use it.

So we probably need another one for when the credential list is empty and we issue a challenge, but we feed the credential list seperately ....